### PR TITLE
fortran: update compiler version

### PR DESCRIPTION
**Pull request by xoviat**

### CODE REVIEW

**Comment by xoviat** [comment]
note migration.yml similar to https://github.com/conda-forge/conda-forge-pinning-feedstock/pull/837/files may be required.

**Review by beckermr — CHANGES_REQUESTED** [review]
This change will need to be discussed with @conda-forge/core and probably requires a migration.

**Comment by xoviat** [comment]
it's been pending for a while now, and wasn't done. I thought I'd shake things loose.

**Comment by isuruf** [comment]
Yes, I don't think there's anything to discuss, but we do need a migration yaml.

**Comment by xoviat** [comment]
Is this okay?

File: recipe/conda_build_config.yaml [hunk 1]
```diff
@@ -26,7 +26,7 @@ fortran_compiler_version:      # [unix or win64]
   - 11                         # [osx and arm64]
   - 9                          # [osx and x86_64]
   - 9                          # [linux]
-  - 5                          # [win64]
+  - 11.0.1                     # [win64]
```
Comment (beckermr) [inline recipe/conda_build_config.yaml:29]: Remove this change until the migration is done.

File: recipe/migrations/flang110.yml [hunk 2]
```diff
@@ -0,0 +1,12 @@
+migrator_ts: 1616270765
+
+__migrator:
+  kind:
+    version
+  migration_number: 1
+  build_number: 1
+  override_cbc_keys:
```
Comment (beckermr) [inline recipe/migrations/flang110.yml:8]:
```suggestion
  longterm: True
  override_cbc_keys:
```

**Comment by beckermr** [comment]
@conda-forge/core Do we want to make an announcement? Should we ship both versions for a while?

**Comment by jakirkham** [comment]
Friendly nudge @conda-forge/core (since this came in on the weekend people may have missed it)

**Comment by xoviat** [comment]
cc @h-vetinari

**Comment by isuruf** [comment]
We still need to add an option to the bot to send this PRs to only packages which use flang on windows. I think right now the bot will send a PR to packages that use fortran on unix, but use `m2w64-gcc` on windows too.

cc @conda-forge/bot

**Comment by h-vetinari** [comment]
Would be awesome if this can be picked up again. 🙃

**Comment by zklaus** [comment]
> We still need to add an option to the bot to send this PRs to only packages which use flang on windows. I think right now the bot will send a PR to packages that use fortran on unix, but use `m2w64-gcc` on windows too.
> 
> cc @conda-forge/bot

@isuruf, any hint on how to do that? I might give it a shot.

**Comment by beckermr** [comment]
How many windows flang packages are there? Maybe we rebuild a few extra ones if it is small?

**Comment by jakirkham** [comment]
How many Fortran packages are there? And how many of those support Windows? Guessing this is probably quite small, but don't really know.

**Comment by zklaus** [comment]
@beckermr, @jakirkham, are you saying that we don't need to modify the bot per @isuruf's suggestion, just because even if this would happen (i.e. some packages get superfluous PRs), those are likely so few that we can deal with them just by manually closing the PRs?

Hence, this migrator can be started as is?

**Comment by beckermr** [comment]
That is my suggestion yes. We'll need everyone to agree before we go that way. I also could be wrong.

**Comment by isuruf** [comment]
No. There's 300+ fortran feedstocks and only about 2 dozen feedstocks using flang. I wouldn't characterize that as `likely so few`.

**Comment by beckermr** [comment]
Ok. So then we’ll need a pr to the bot to avoid rebuilding about 30 feedstocks out of 300. I don’t think I’ll have time to do this but I’m happy to review!

**Comment by isuruf** [comment]
> So then we’ll need a pr to the bot to avoid rebuilding about 30 feedstocks out of 300

No, it's the other way around. We have to avoid rebuilding 270 out of 300.

**Comment by zklaus** [comment]
Gotcha. Do I understand correctly that this needs a PR to change [`auto_tick.py`](https://github.com/regro/cf-scripts/blob/master/conda_forge_tick/auto_tick.py)?

@isuruf, did you have something general in mind, like an addition to the yaml format for the migrators that allows filtering on the graph, or rather something specific, more like a [custom migrator](https://regro.github.io/cf-scripts/migrators.html#custom-migrations)?

**Comment by beckermr** [comment]
> No, it's the other way around. We have to avoid rebuilding 270 out of 300.

Ahh. I am still not terribly worried about that. 

As for the bot, we likely want to add a platform filtering option to the standard yaml migratory here: https://github.com/regro/cf-scripts/blob/master/conda_forge_tick/migrators/migration_yaml.py#L113

Then we can list the platforms to target and avoid a custom migrator.

**Comment by zklaus** [comment]
I'll see what I can do. But if anybody else gets to it before me, I won't be mad :stuck_out_tongue:.

**Comment by h-vetinari** [comment]
@zklaus any news on your efforts with this? 🙃

**Comment by zklaus** [comment]
Sorry, @h-vetinari, I have been taken up by other things. I am not sure when I will manage to pick this up in earnest.

**Comment by h-vetinari** [comment]
With llvm-flang hopefully not toooooo far out, I gave this a shot in https://github.com/regro/cf-scripts/pull/1655. Feedback & advice appreciated!

File: recipe/migrations/flang110.yml [hunk 3]
```diff
@@ -0,0 +1,13 @@
+migrator_ts: 1616270765
+
+__migrator:
+  kind:
+    version
+  migration_number: 1
+  build_number: 1
+  longterm: True
```
Comment (h-vetinari) [inline recipe/migrations/flang110.yml:8]:
```suggestion
  longterm: True
  platform_allowlist:
    - win-64
```

**Review by h-vetinari — COMMENTED** [review]
@xoviat, not sure if you're still around? 😅 

If yes, could you try to add the following (based on https://github.com/regro/cf-scripts/pull/1655)? Since we couldn't write tests I'm a bit apprehensive if it'll work right away, but we can always pause & debug the migration.

PS. Not sure if relevant; clang 11.0 had an ABI break and 11.1 was released to fix it. I had an open [PR](https://github.com/conda-forge/flang-feedstock/pull/24) to rebuild flang for that, but this ran into some weird build tooling errors I wasn't able to debug at the time.

**Comment by h-vetinari** [comment]
@isuruf, are your concerns about the migration now addressed? I believe the very old default flang version is the reason why the compilers package is having resolution problems: https://github.com/conda-forge/compilers-feedstock/issues/58

**Comment by h-vetinari** [comment]
As an update to anyone subscribed here - we discussed this in a core meeting a few months [ago](https://github.com/conda-forge/conda-forge.github.io/blob/main/src/orga/minutes/2023-06-28.md), and the status was that the "updated" flang 11 builds based on classic-flang have no support (we also haven't heard from @xoviat for a long time), and so we decided to wait for llvm-flang to materialise.

We now have flang 17 live in conda-forge, but it's still somewhat experimental, so the earliest to retry this realistically is for flang 18 early next year.

**Comment by h-vetinari** [comment]
Another update: flang 18 didn't work for compiling OpenBLAS on windows, but with the in-progress flang 19 (ETA ~September), things are looking much [better](https://github.com/conda-forge/openblas-feedstock/pull/115#issuecomment-2215517076).

**Comment by h-vetinari** [comment]
This is still a draft, but it's looking like flang 19 will finally be ready on windows to compile most of our feedstocks. Already tested successfully: OpenBLAS, SciPy and a few others in progress.

**Comment by h-vetinari** [comment]
Finally the rename from flang-new to flang landed: https://github.com/llvm/llvm-project/commit/06eb10dadfaeaadc5d0d95d38bea4bfb5253e077 🥳 

I've tried backporting the patch to LLVM 19, and aside from some trivial conflicts, it looks like we could do it, though I think I'd prefer to leave that for LLVM 20. Let me know if someone thinks differently...

**Comment by h-vetinari** [comment]
Alright, let's get this party started!

**Review by beckermr — COMMENTED** [review]
Well I totally missed this, but the migration files need to end in `.yaml`. We should add a test and/or hook in pre-commit for this.

**Pull request merged**